### PR TITLE
Fix setup upgrade command re enables caches

### LINKED ISSUE #28186: Running bin/magento setup:upgrade re-enables layout, block_html and full_page caches
### Preconditions (*)
1.Magento 2.4-develop

### Steps to reproduce (*)
1. Disable the caches by running: `bin/magento cache:disable block_html full_page layout`
2. Run: `bin/magento setup:upgrade`

### Expected result (*)
1. After setup:upgrade completes execution, the `block_html`, `full_page`, and `layout` caches are still disabled

### Actual result (*)
1. An "enabling caches" message appears in the CLI output of setup:upgrade:
```
Enabling caches:
Current status:
Array
(
    [layout] => 1
    [block_html] => 1
    [full_page] => 1
)
```
2. Those caches are re-enabled:
```
magento235$ bin/magento cache:status
Current status:
                        config: 1
                        layout: 1
                    block_html: 1
                   collections: 1
                    reflection: 1
                        db_ddl: 1
               compiled_config: 1
                           eav: 1
         customer_notification: 1
            config_integration: 1
        config_integration_api: 1
                google_product: 1
                     full_page: 1
             config_webservice: 1
                     translate: 1
                        vertex: 1
```

### PR DESCRIPTION
**Pull request by AbdulRahmanAbouzaid**
### Description (*)

Fix re-enables all frontend caches after running setup:upgrade command including the disabled one BY Return them back to their old state before running setup:upgrade.

### Fixed Issues (if relevant)

1. Fixes magento/magento2#28186

### Manual testing scenarios (*)

1.Disable one of/all frontend caches [ layout, block_html and full_page] (for example: run `bin/magento cache:disable layout`).
2. run `bin/magento setup:upgrade`.
3. Run `bin/magento cache:status` and you can find that layout cache disabled.

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by AbdulRahmanAbouzaid** [comment]
@magento run all tests

**Comment by AbdulRahmanAbouzaid** [comment]
@magento run Unit Tests, WebAPI Tests, Sample Data Tests EE, Sample Data Tests B2B, Integration Tests, Functional Tests EE, Functional Tests CE, Functional Tests B2B

**Comment by AbdulRahmanAbouzaid** [comment]
@magento run all tests

**Comment by AbdulRahmanAbouzaid** [comment]
@magento run Static Tests

**Comment by AbdulRahmanAbouzaid** [comment]
@magento run Integration Tests

**Comment by AbdulRahmanAbouzaid** [comment]
@magento run Functional Tests B2B

**Comment by AbdulRahmanAbouzaid** [comment]
@magento run all tests

**Comment by VladimirZaets** [comment]
@magento run all tests

**Review by VladimirZaets — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @VladimirZaets, thank you for the review.
[ENGCOM-8029](https://jira.corp.magento.com/browse/ENGCOM-8029) has been created to process this Pull Request
:eight_spoked_asterisk: @VladimirZaets, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

**Comment by engcom-Alfa** [comment]
@VladimirZaets Could you put an appropriate label for test coverage?

**Comment by ihor-sviziev** [comment]
@engcom-Alfa I believe it requires automatic test, but not sure how to test CLI commands correctly

File: setup/src/Magento/Setup/Model/Installer.php [hunk 1]
```diff
@@ -947,6 +959,10 @@ public function installDataFixtures(array $request = [])
         $this->handleDBSchemaData($setup, 'data', $request);
         $this->log->log('Enabling caches:');
         $this->updateCaches(true, $frontendCaches);
+        if ($keepCacheStatuses && !empty($disabledCaches)) {
+            $this->log->log('Disabling pre-disabled caches:');
+            $this->updateCaches(false, array_keys($disabledCaches));
+        }
```
Comment (ihor-sviziev) [inline setup/src/Magento/Setup/Model/Installer.php]: It seems to me really bad to enable caches and disable them again. I think it's better to calculate what caches need to be re-enabled, and enable only needed

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]

**Comment by ihor-sviziev** [comment]
@AbdulRahmanAbouzaid could you also cover your changes with some kind of test?

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@magento run all tests

File: setup/src/Magento/Setup/Model/Installer.php (hunk 1 above)
Comment (ihor-sviziev) [reply setup/src/Magento/Setup/Model/Installer.php]: Just fixed it myself

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by AbdulRahmanAbouzaid** [comment]
@ihor-sviziev thanks for your updates. Is there any other changes should be made?

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Hi @AbdulRahmanAbouzaid,
Could you cover your changes with any type of automated tests in order to prevent appearing this issue in future?

**Comment by AbdulRahmanAbouzaid** [comment]
@ihor-sviziev I'll check this

**Comment by engcom-Charlie** [comment]
Hi @AbdulRahmanAbouzaid, do you have any updates about test coverage?
Thank you.

**Comment by gabrieldagama** [comment]
The risk was set to `medium` due to: the suggested changes may affect a critical area.

**Comment by engcom-Charlie** [comment]
@AbdulRahmanAbouzaid I'll try to continue with your PR.

**Comment by AbdulRahmanAbouzaid** [comment]
@engcom-Charlie Thank you. 
Sorry for the late reply. I tried to cover this with test but i couldn't

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Review by ihor-sviziev — APPROVED** [review]
✔ Approved.

Failing tests looks not related to changes form this PR.

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-8029](https://jira.corp.magento.com/browse/ENGCOM-8029) has been created to process this Pull Request

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by engcom-Bravo** [comment]
:heavy_check_mark: QA Passed

The cache status has not been changed after performing Upgrade

### Manual testing scenario
1. Disable _block_html_ _full_page_ _layout_ using the following command
```bin/magento cache:disable block_html full_page layout```
2. Run ```bin/magento cache:status``` and assert that following are set to _0_
```layout: 0```
```block_html: 0```
```full_page: 0 ```
3. Upgrade magento instance using ```bin/magento setup:upgrade```
4. Check the cache status using  ```bin/magento cache:status```

### Actual Result
The previously disabled caches are still in _Disabled_ status (0)
```
Current status:
                        config: 1
                        layout: 0
                    block_html: 0
                   collections: 1
                    reflection: 1
                        db_ddl: 1
               compiled_config: 1
                           eav: 1
         customer_notification: 1
            config_integration: 1
        config_integration_api: 1
                     full_page: 0
             config_webservice: 1
                     translate: 1
```
### Before switching to PR 
```
Current status:
                        config: 1
                        layout: 1
                    block_html: 1
                   collections: 1
                    reflection: 1
                        db_ddl: 1
               compiled_config: 1
                           eav: 1
         customer_notification: 1
            config_integration: 1
        config_integration_api: 1
                     full_page: 1
             config_webservice: 1
                     translate: 1
```

**Pull request merged**